### PR TITLE
[WIP] Integrating DISK

### PR DESCRIPTION
**Pull request by jatentaki**
#### Changes
First step in upstreaming DISK. This is a working prototype, a lot of unnecessary code can be cut out, linting etc was not done.

CC @ducha-aiki 

#### Type of change
- [ ] 🧪 Tests Cases
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by ducha-aiki** [comment]
DISK makes use of `np.ndarray(dtype=object)` as a hack around not having JaggedTensors (having different number of keypoints per image).

Right now we don't support that: we either do per-image as for `KeyNet`, or expect same number of keypoints per image as for `ScaleSpaceDetector`.


@jatentaki: 

> TLDR: numpy is pytorch's dependency, so removing it won't make kornia "leaner". Also, removing is somewhat complicated (but not impossible). In terms of user-facing API, we can just call .tolist() and I have 0 issues with that (in factor support the motion).
And if you're returning a list, why not a ndarray which can be arbitrarily reshaped (which is useful internally)?

So, possible action points are: 

(a). lift the pytorch-only requirement and use numpy directly. This may affect maintainability, but also allow new usages

(b) write our own write around, e.g. in form of list of loftr-style: output results `[N, dim]`, idx of size `[N]`. This would solve the inference thing and enhance kornia local features, but disallow to add DISK training in future. 

(c) something else?

@edgarriba @johnnv1 @shijianjian thoughts?

File: kornia/feature/disk/detector.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        ).permute(
+            0, 2, 3, 1
+        )  # -> bhw2
+
+        keypoints = []
+        for i in range(B):
+            mask = accept_mask[i]
+            keypoints.append(Keypoints(xys[i][mask], logp[i][mask]))
+
+        return np.array(keypoints, dtype=object)
```
Comment (edgarriba) [inline kornia/feature/disk/detector.py:124]: why don't return just the list?

Comment (ducha-aiki) [reply kornia/feature/disk/detector.py:124]: @edgarriba see the discussion in PR

File: kornia/feature/disk/detector.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+        # tile the heatmap into [window x window] tiles and pass it to
+        # the categorical distribution.
+        heatmap_tiled = self._tile(heatmap).squeeze(1)
+        proposals, accept_mask, logp = point_distribution(heatmap_tiled)
+
+        # create a grid of xy coordinates and tile it as well
+        cgrid = torch.stack(
+            torch.meshgrid(torch.arange(H, device=dev), torch.arange(W, device=dev), indexing='ij')[::-1], dim=0
+        ).unsqueeze(0)
```
Comment (edgarriba) [inline kornia/feature/disk/detector.py:106]: we have a utility function `create_meshgrid` for this

File: kornia/feature/disk/detector.py (hunk 1 above)
Comment (jatentaki) [reply kornia/feature/disk/detector.py:124]: It doesn't matter until you train where you have [N_scenes, N_view_per_scene] and flatten that for Unet forward then unflatten for computing loss. But at that point you have different number of keypoints so can't use regular tensors

File: kornia/feature/disk/detector.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        for i in range(B):
+            mask = accept_mask[i]
+            keypoints.append(Keypoints(xys[i][mask], logp[i][mask]))
+
+        return np.array(keypoints, dtype=object)
+
+    def nms(self, heatmap: Tensor, n: Optional[int] = None, **kwargs) -> np.ndarray:
+        """Inference-time nms-based detection protocol."""
+        heatmap = heatmap.squeeze(1)
+        nmsed = nms(heatmap, **kwargs)
```
Comment (edgarriba) [inline kornia/feature/disk/detector.py:129]: i don't see a reason why this cannot be just `list[Keypoint]` since what you only do is random access to the list
+ suggest to migrate typing to `from __future__ import annotate`

File: kornia/feature/disk/detector.py (hunk 2 above)
Comment (jatentaki) [reply kornia/feature/disk/detector.py:106]: Up to the user, but can be. We can cache the last grid and just update if the next incoming image has different shape

File: kornia/feature/disk/disk.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        """allowed values for `algorithm`:
+
+        * rng
+        * nms
+        """
+
+        B = images.shape[0]
+        heatmaps, descriptors = self.heatmap_and_dense_descriptors(images)
+
+        keypoints: Sequence[Keypoints] = {'rng': self.detector.sample, 'nms': self.detector.nms}[algorithm](
```
Comment (edgarriba) [inline kornia/feature/disk/disk.py:60]: with sequence and not list

Comment (edgarriba) [reply kornia/feature/disk/disk.py:60]: not very readable code; suggest to have an instance method
```python
def compute_keypoints(self, heatmap: Tensor, algorithm: str) -> list[Keypoints]:
   KORNIA_CHECK(algorithm in ("rng, "nms")
    if algorithm == "rng":
        return  self.detector.sample(heatmaps)
    return self.detector.nms(heatmap)
```
this is more testable/predictable

File: kornia/feature/disk/structs.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    def n(self):
+        return self.keypoints.shape[0]
+
+    @property
+    def device(self):
+        return self.keypoints.device
+
+    @property
+    def x(self):
+        return self.keypoints[:, 0]
```
Comment (edgarriba) [inline kornia/feature/disk/structs.py:28]:
```suggestion
        return self.keypoints[..., 0]
```
does it make sense like this?

File: kornia/feature/disk/detector.py (hunk 1 above)
Comment (edgarriba) [reply kornia/feature/disk/detector.py:124]: what feature bring numpy here that is required later ? as Im seeing the usage later you only random access, which is solvable by return just a list (or lists if needed) of that `Keypoints` data structure

Comment (jatentaki) [reply kornia/feature/disk/detector.py:124]: You don't see anything because you're looking only at inference code. The question is if we want to have the training code in there as well. And I'd say no, but @ducha-aiki seems to think it could be useful.

File: kornia/feature/disk/detector.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+        keypoints = []
+        for b in range(heatmap.shape[0]):
+            yx = nmsed[b].nonzero(as_tuple=False)
+            logp = heatmap[b][nmsed[b]]
+            xy = torch.flip(yx, (1,))
```
[the remark this answers is not in the document]
Comment (jatentaki) [reply kornia/feature/disk/detector.py:135]: What does that mean? I don't think this code will ever work with numpy backend (unless we do a heavy rewrite).

File: kornia/feature/disk/structs.py (hunk 5 above)
Comment (jatentaki) [reply kornia/feature/disk/structs.py:28]: It wouldn't be a bug, but the keypoints struct always holds a 2-dimensional tensor (i.e. not batched) and writing `...` implies it could be more than 2d which is not true.

**Comment by jatentaki** [comment]
Overall I think integrating the training code would be a pain because it depends on specific data format, which comes from specific data loaders, which are implemented for specific datasets, and so on and so forth. So I think we can just ditch the idea of including training code in Kornia and refer the user to my original repo which does not need to adhere to this strict levels of maintainability and version correctness. The user simply uses `kornia.feature.disk` with pretrained weights and if they find they like the algorithm but need to fine-tune it, they clone my original repo, train there, and load that checkpoint in their `kornia`-using code.

**Comment by jatentaki** [comment]
To be clear: the above suggestion would remove any remaining reasons to use `numpy`, clearing that issue.

**Comment by ducha-aiki** [comment]
@jatentaki OK, then let's do that. And if we want to use some kind of training, then it would be a separate repo like yours.

**Comment by edgarriba** [comment]
agree on the training side

File: kornia/feature/disk/disk.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                raise
+
+        assert unet_output.shape[1] == self.desc_dim + 1
+
+        descriptors = unet_output[:, : self.desc_dim]
+        heatmaps = unet_output[:, self.desc_dim :]
+
+        return heatmaps, descriptors
+
+    def detect(self, images: Tensor, algorithm: Literal['rng', 'nms'] = 'nms', **kwargs) -> Sequence[DISKFeatures]:
```
[the remark this answers is not in the document]
Comment (johnnv1) [reply kornia/feature/disk/disk.py:50]: not that I know of

[the remark this answers is not in the document]
Comment (jatentaki) [reply kornia/feature/disk/disk.py]: replaced with `algorithm: str`

File: kornia/feature/disk/detector.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jatentaki) [reply kornia/feature/disk/detector.py]: Did the change, still curious :)

File: kornia/feature/disk/detector.py (hunk 2 above)
Comment (jatentaki) [reply kornia/feature/disk/detector.py]: So far replaced my handrolled meshgrid with `kornia.utils.create_meshgrid`

File: kornia/feature/disk/detector.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    accept_samples = accept_dist.sample()
+    accept_logp = accept_dist.log_prob(accept_samples)
+    accept_mask = accept_samples == 1.0
+
+    logp = proposal_logp + accept_logp
+
+    return proposals, accept_mask, logp
+
+
+def nms(signal: Tensor, window_size=5, cutoff=0.0) -> Tensor:
```
Comment (ducha-aiki) [inline kornia/feature/disk/detector.py:48]: @jatentaki wouldn't `kornia.geometry.subpix.nms2d` work here?
https://github.com/kornia/kornia/blob/master/kornia/geometry/subpix/nms.py#L27

Comment (jatentaki) [reply kornia/feature/disk/detector.py:48]: True :) though my implementation seems more efficient - you're using conv2d with `eye` kernel and then max, while I use a single call to `max_pool2d`

**Comment by jatentaki** [comment]
Question: do we want to keep the randomly sampled keypoints, as opposed to NMS? I mostly left it in DISK code to enable people to compare the gap between training regime and inference, but otherwise we know NMS is straight up better. I vote no (don't keep rng).

**Comment by ducha-aiki** [comment]
Let's go for the minimalistic implementation.

**Comment by jatentaki** [comment]
Ok. What is left to do?

1. Add typing hints to the `unets` submodule
2. Add documentation
3. More tests? Can we test the semantics of the network somehow? Do you have experience in writing more than just smoke tests to detect potential breakage?
4. Relax the multiple-of-16 requirement? It's easy enough to lift it and keep the same behavior for multiple-of-16 inputs, but I have no clue how well the network would perform outside this regime, so users could experience silent performance losses.
5. Do you want more inline comments?
6. Anything else?

**Comment by ducha-aiki** [comment]
@jatentaki 

 3. I'd go for running on smallest image possible - like 128x128, saving the outputs and testing against them in unit test, thus we will know if smth breaks.
 4. Let's first release it as is, and then with some tests add automatic input padding?
 
The rest is enough, I think

File: kornia/feature/disk/unets/unet.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+import torch.nn as nn
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/feature/disk/unets/unet.py:1]: No, because this particular instance has this some input bug, as we discussed

**Comment by ducha-aiki** [comment]
@jatentaki sure, will do. 
Also, let's remove the example from here and more it to the https://github.com/kornia/tutorials to keep main repo small-ish.

File: kornia/feature/disk/detector.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+from typing import List, Optional
```
Comment (johnnv1) [inline kornia/feature/disk/detector.py:1]:
```suggestion
from __future__ import annotations
```

File: kornia/feature/disk/disk.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+from kornia.core import Tensor
+from kornia.utils.helpers import map_location_to_cpu
+
+from .detector import heatmap_to_keypoints
+from .structs import DISKFeatures
+from .unets import Unet
+
+
+class DISK(torch.nn.Module):
```
Comment (johnnv1) [inline kornia/feature/disk/disk.py:13]: `from kornia.core import Module`
```suggestion
class DISK(Module):
```

File: kornia/feature/disk/disk.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        try:
+            unet_output = self.unet(images)
+        except RuntimeError as e:
+            if 'Trying to downsample' in str(e):
+                raise ValueError(
+                    f'The current implementation of U-Net requires image size to '
+                    f'be divisible by 16 (got {images.shape[2:]}).'
+                )
+            else:
+                raise
```
Comment (johnnv1) [inline kornia/feature/disk/disk.py:51]: we can't test it before forward unet? to avoid a unecessary try-except here

File: kornia/feature/disk/disk.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
+from typing import List, Optional, Tuple
```
Comment (johnnv1) [inline kornia/feature/disk/disk.py:1]:
```suggestion
from __future__ import annotations
```

File: kornia/feature/disk/disk.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        if checkpoint == 'depth':
+            pretrained_dict = torch.hub.load_state_dict_from_url(
+                'https://raw.githubusercontent.com/cvlab-epfl/disk/master/depth-save.pth',
+                map_location=map_location_to_cpu,
+            )
+        elif checkpoint == 'epipolar':
+            pretrained_dict = torch.hub.load_state_dict_from_url(
+                'https://raw.githubusercontent.com/cvlab-epfl/disk/master/epipolar-save.pth',
+                map_location=map_location_to_cpu,
+            )
```
Comment (johnnv1) [inline kornia/feature/disk/disk.py:113]: load direct to `device` here didn't work? e.g to avoid load on cpu and move to gpu

File: kornia/feature/disk/structs.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    varying number of keypoints per image, `DISKFeatures` is not batched.
+
+    Args:
+        keypoints: Tensor of shape :math:`(N, 2)`, where :math:`N` is the number of keypoints.
+        descriptors: Tensor of shape :math:`(N, 128)`.
+        detection_scores: Tensor of shape :math:`(N,)` where the detection score can be interpreted as
+        the log-probability of keeping a keypoint after it has been proposed (see the paper,
+        section Method->Feature distribution for details).
+    """
+    keypoints: Tensor
```
Comment (johnnv1) [inline kornia/feature/disk/structs.py:21]: i didn't get why we had a dataclass for keypoints, and here is just a Tensor. Can't be that struct or https://github.com/kornia/kornia/blob/master/kornia/geometry/keypoints.py#L13?

File: kornia/feature/disk/structs.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    def to(self, *args, **kwargs):
+        return DISKFeatures(
+            self.keypoints.to(*args, **kwargs),
+            self.descriptors.to(*args, **kwargs),
+            self.detection_scores.to(*args, **kwargs),
+        )
+
+
+@dataclass
+class Keypoints:
```
Comment (johnnv1) [inline kornia/feature/disk/structs.py:50]: I don't have a name suggestion, but I think we could avoid that name to avoid confusion with https://github.com/kornia/kornia/blob/master/kornia/geometry/keypoints.py#L13

File: kornia/feature/disk/unets/blocks.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from kornia.core import Tensor
+
+from .utils import cut_to_match, size_is_pow2
+
+
+class TrivialUpsample(nn.Module):
```
Comment (johnnv1) [inline kornia/feature/disk/unets/blocks.py:10]: from kornia.core
```suggestion
class TrivialUpsample(Module):
```

same for the others

File: test/feature/test_disk.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import torch
+
+from kornia.feature.disk import DISK
+
+
+class TestDisk:
```
Comment (johnnv1) [inline test/feature/test_disk.py:6]: these tests can't be with smaller inputs?

File: kornia/feature/disk/unets/unet.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (johnnv1) [reply kornia/feature/disk/unets/unet.py:1]: Sugestion: rename this submodule to `_unets` since its just to be internally used for the DISK module

File: kornia/feature/disk/detector.py (hunk 8 above)
Comment (johnnv1) [reply kornia/feature/disk/detector.py]: I think this worth at least open a issue to map these nms, and try to unify into the more efficient case  -- with this pr, i think will be at least the third nms inside kornia :)

Comment (ducha-aiki) [reply kornia/feature/disk/detector.py]: @johnnv1 they also work a bit differently. MaxPool-based (in DISK) returns 1 for torch.ones(3, 3). Conv-based (in subpix) allows only strict maximum

File: kornia/feature/disk/structs.py (hunk 16 above)
Comment (ducha-aiki) [reply kornia/feature/disk/structs.py:50]: Well, they both are keypoints, and DISK keypoints are not really LAFs (kornia structure for local features), as they lack scale. Renaming them probably would cause even more confusion, so I'd keep them as is.

File: kornia/feature/disk/detector.py (hunk 10 above)
Comment (jatentaki) [reply kornia/feature/disk/detector.py:1]: What does this do?

File: kornia/feature/disk/disk.py (hunk 11 above)
Comment (jatentaki) [reply kornia/feature/disk/disk.py:13]: What is the difference?

File: kornia/feature/disk/disk.py (hunk 12 above)
Comment (jatentaki) [reply kornia/feature/disk/disk.py:51]: Good point.

File: kornia/feature/disk/structs.py (hunk 16 above)
Comment (jatentaki) [reply kornia/feature/disk/structs.py:50]: Above, plus this is an internal data structure that the user should never see (it will be merged with descriptors and turned into `DISKFeatures` before the API boundary).

File: kornia/feature/disk/structs.py (hunk 15 above)
Comment (jatentaki) [reply kornia/feature/disk/structs.py:21]: I wasn't aware of `geometry.Keypoints`. Do any other local feature modules use it? It appears that `geometry.Keypoints` is meant as a batchable data container (since it's mostly referenced in data augmentation modules). Why not `disk.Keypoints`? Because I think it's an unnecessary level of indirection, `DISKFeatures` is not cluttered with just 3 members, so merging two of them into a struct of their own won't help usability.

Comment (jatentaki) [reply kornia/feature/disk/structs.py:21]: @edgarriba do I understand correctly that you're in favor of keeping it as a `Tensor`? I am - `geometry.Keypoints` would be compatible, it would provide some extra functionality (affine transformations) but I don't think this is very useful to the average user and increases the mental burden of searching through documentation.

File: kornia/feature/disk/disk.py (hunk 11 above)
Comment (johnnv1) [reply kornia/feature/disk/disk.py:13]: We are exposing some things on the `core' to facilitate manteability and try to do something in the future

File: test/feature/test_disk.py [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+        inp = torch.ones(1, 3, 256, 256, device=device)
+        output = disk(inp)
+        assert output is not None
+
+    def test_smoke_epipolar(self, device):
+        disk: DISK = DISK.from_pretrained(checkpoint='epipolar', device=device)
+        inp = torch.ones(1, 3, 256, 256, device=device)
+        output = disk(inp)
+        assert output is not None
+
```
Comment (johnnv1) [inline test/feature/test_disk.py:18]: btw, would be nice if we can add a `dynamo` test here. E.g https://github.com/kornia/kornia/blob/master/test/filters/test_median.py#L82

File: test/feature/test_disk.py [hunk 20]
```diff
@@ -0,0 +1,25 @@
+import torch
+
+from kornia.feature.disk import DISK
+
+
+class TestDisk:
+    def test_smoke_depth(self, device):
+        disk: DISK = DISK.from_pretrained(checkpoint='depth', device=device)
```
Comment (johnnv1) [inline test/feature/test_disk.py:8]: can this smoke test be only `DISK` without loading the weights?

**Comment by jatentaki** [comment]
Resolved most of the above. I kept one smoketest with `from_pretrained` because this is a piece of code that could be buggy in itself. I'm leaving dynamo to you because again I don't know at what granularity we want to assert.

File: test/feature/test_disk.py [hunk 21]
```diff
@@ -1,25 +1,39 @@
...
 
-from kornia.feature.disk import DISK
+from kornia.feature.disk import DISK, DISKFeatures
 
 
 class TestDisk:
-    def test_smoke_depth(self, device):
-        disk: DISK = DISK.from_pretrained(checkpoint='depth', device=device)
-        inp = torch.ones(1, 3, 256, 256, device=device)
+    def test_smoke(self, device):
```
Comment (edgarriba) [inline test/feature/test_disk.py:8]:
```suggestion
    def test_smoke(self, dtype, device):
```

Comment (jatentaki) [reply test/feature/test_disk.py:8]: What would be the semantics? That we cast the input (and the network un-casts it)? Or that both inputs and network are cast and the whole computation is in `dtype`?

Comment (jatentaki) [reply test/feature/test_disk.py]: Done

File: test/feature/test_disk.py [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+class TestDisk:
+    def test_smoke(self, dtype, device):
+        disk = DISK().to(device, dtype)
+        inp = torch.ones(1, 3, 64, 64, device=device, dtype=dtype)
+        output = disk(inp)
+        assert isinstance(output, list)
+        assert len(output) == 1
+        for element in output:
+            assert isinstance(element, DISKFeatures)
```
Comment (johnnv1) [inline test/feature/test_disk.py:15]: I think, this should work
```suggestion
        assert all(isinstance(element, DISKFeatures) for element in output)
```

Same for the others tests

Comment (jatentaki) [reply test/feature/test_disk.py]: Done

File: test/feature/test_disk.py (hunk 19 above)
Comment (ducha-aiki) [reply test/feature/test_disk.py:36]: @johnnv1 @jatentaki good point about dynamo, let's do it as a separate PR, as we are time-constrained to get DISK into kornia before IMC2023 starts.

**Review by ducha-aiki — APPROVED** [review]

**Comment by jatentaki** [comment]
Great thanks! A note in change log?
